src/osmet/mod.rs

```diff
@@ -0,0 +1,579 @@
+        end_offset,
+        mappings,
+    })
+}
+
+/// Writes the disk image, with the extents for which we have mappings for "punched out".
+fn write_punched_image_to_file(
+    block_device: &Path,
+    partitions: &[OsmetPartition],
+) -> Result<(File, u64)> {
+    let mut xz_tmpf = XzEncoder::new(tempfile::tempfile()?, 9);
+
+    let mut dev = OpenOptions::new()
+        .read(true)
+        .open(&block_device)
+        .chain_err(|| format!("opening {:?}", block_device))?;
+
+    let total_bytes_punched = write_punched_image(&mut dev, &mut xz_tmpf, partitions)?;
+
+    xz_tmpf
+        .try_finish()
+        .chain_err(|| "while finishing xz stream")?;
+
+    // sanity check that the number of bytes written + punched match up with block device size
+    let blksize = get_block_size(&dev)?;
+    let total_bytes_written = xz_tmpf.total_in();
+    if total_bytes_written + total_bytes_punched != blksize.get() {
+        bail!(
+            "bytes written + bytes punched != block device size: {} + {} vs {}",
+            total_bytes_written,
+            total_bytes_punched,
+            blksize
+        );
+    }
+
+    println!("Total bytes skipped: {}", total_bytes_punched);
+    println!("Total bytes written: {}", total_bytes_written);
+    println!("Total bytes written (compressed): {}", xz_tmpf.total_out());
+
+    let mut tmpf = xz_tmpf.finish()?;
+    tmpf.seek(SeekFrom::Start(0))?;
+
+    Ok((tmpf, blksize.get()))
+}
+
+fn write_punched_image(
+    dev: &mut File,
+    w: &mut impl Write,
+    partitions: &[OsmetPartition],
+) -> Result<u64> {
+    let mut buf = [0u8; 8192];
+
+    let mut cursor: u64 = 0;
+    let mut total_bytes_punched: u64 = 0;
+    for partition in partitions {
+        // first copy everything up to the start of the partition
+        assert!(partition.start_offset >= cursor);
+        copy_exactly_n(dev, w, partition.start_offset - cursor, &mut buf)?;
+        total_bytes_punched += write_punched_image_partition(dev, w, partition, &mut buf)?;
+        cursor = partition.end_offset;
+    }
+
+    // and finally write out the remainder of the disk
+    io::copy(dev, w)?;
+
+    Ok(total_bytes_punched)
+}
+
+fn write_punched_image_partition(
+    dev: &mut File,
+    w: &mut impl Write,
+    partition: &OsmetPartition,
+    buf: &mut [u8],
+) -> Result<u64> {
+    let mut total_bytes_punched = 0;
+
+    // and this is where the real fun begins!
+    let mut cursor = partition.start_offset;
+    for mapping in partition.mappings.iter() {
+        // make offset relative to start of disk, not partition
+        let extent_start = mapping.extent.physical + partition.start_offset;
+        assert!(extent_start >= cursor);
+        if cursor < extent_start {
+            cursor += copy_exactly_n(dev, w, extent_start - cursor, buf)
+                .chain_err(|| "while writing in between extents")?;
+        }
+
+        // this is the crucial space-saving step; we skip over the extent we have a mapping for
```

Comment:
I've standardized on the following terms throughout the osmet codebase:
- the "unpacked" image refers to the fully blown up metal image
- extents for which we already have a mapping are "skipped" (instead of "punched")
- the "packed" image refers to the metal image with all the extents for which we already have a mapping skipped
- the "xzpacked" image is the packed image compressed with xz

Reply:
Could we get the above in a comment?